### ISSUE TITLE
[Enhancement] [Kobo] one-step installer 

### ISSUE THREAD

**Issue by pazos** [koreader/koreader#1996]
Project: integrate a (minimal)launcher based on inotify (like fmon) and koreader in KoboRoot.tgz
Goals: provide an alternative for new users, or users that don't like to install a full blown launcher, like KSM.

launcher is coded in c, it just wait for IN_OPEN events on /mnt/onboard/koreader.png
when our target is opened a helper script is launched. The helper script finds a match for file:///mnt/onboard/koreader.png in nickel's db and runs koreader just in that case.

I submit a dirty way of doing this, as an example:
[kmon-sample.zip](https://github.com/koreader/koreader/files/239707/kmon-sample.zip)

What is your opinion?

**Comment by pazos** [comment]
There are some typos in the code, but I hope the concept is clear :)

**Comment by NiLuJe** [comment]
Just for the record, since I've never used KSM nor fmon:

* PNG files dropped somewhere in onboard are parsed by Nickel and creates dedicated tiles?

* Can't remember if the FW ships w/ the sqlite3 client, but doing a proper SQL query (either via the CLI client, or merging everything in the C code end using libsqlite3 directly) seems "cleaner", as a goal ;).

**Comment by pazos** [comment]
I don't understand SQL at all, but the PNG seems to be referenced in some places:
`grep file:///mnt/onboard/koreader.png /mnt/onboard/.kobo/KoboReader.sqlite | wc -l`
= 17

`grep faulkner /mnt/onboard/.kobo/KoboReader.sqlite | wc -l`  <-- just one ebook
= 179

**Comment by NiLuJe** [comment]
```
sqlite3 /mnt/onboard/.kobo/KoboReader.sqlite "SELECT EXISTS(SELECT 1 FROM content WHERE ContentID = 'file:///mnt/onboard/koreader.png' AND ContentType = '6');"
```

Returns 1 if found (i.e., in the Library), 0 otherwise.

**Comment by NiLuJe** [comment]
Note that, yep, my memory wasn't wrong, the Kobos do NOT ship w/ the sqlite3 client...

```
┌─(ROOT@(none):pts/0)───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(/)─┐
└─(0.54:53%:06:32:86%:#)── ls -lash $(which sqlite3)                                                                                                                                                                                                        ──(Thu, Apr 28)─┘
     0 lrwxrwxrwx    1 root     root          39 Mar  2 23:19 /usr/bin/sqlite3 -> /mnt/onboard/.niluje/python/bin/sqlite3
```

**Comment by NiLuJe** [comment]
And to answer my initial question, since I've now checked: yeah, PNGs create both a tile and an entry in the Library....

(... and are wrongly assigned the application/x-cbz mimetype, but, hey... :D).

**Comment by NiLuJe** [comment]
TL;DR: I'm all for it!

We have [http://trac.ak-team.com/trac/browser/niluje/Configs/trunk/Kindle/Kobo_Hacks/KoboStuff/src/etc/udev/rules.d/99-stuff.rules](non-intrusive) ways to launch daemons at boot, so, yeeep, I like it.

**Comment by NiLuJe** [comment]
And whether opened from a tile of the Library, the string of inotify events is the same:

```
/mnt/onboard/koreader.png OPEN 
/mnt/onboard/koreader.png ACCESS 
/mnt/onboard/koreader.png ACCESS 
/mnt/onboard/koreader.png ACCESS 
/mnt/onboard/koreader.png ACCESS 
/mnt/onboard/koreader.png ACCESS 
/mnt/onboard/koreader.png ACCESS 
/mnt/onboard/koreader.png ACCESS 
/mnt/onboard/koreader.png ACCESS 
/mnt/onboard/koreader.png ACCESS 
/mnt/onboard/koreader.png ACCESS 
/mnt/onboard/koreader.png ACCESS 
/mnt/onboard/koreader.png ACCESS 
/mnt/onboard/koreader.png ACCESS 
/mnt/onboard/koreader.png ACCESS 
/mnt/onboard/koreader.png ACCESS 
/mnt/onboard/koreader.png CLOSE_NOWRITE,CLOSE
```

**Comment by NiLuJe** [comment]
The [SQLite API](https://sqlite.org/quickstart.html) looks dead simple and easy to use, so i'm all for doing everything in C and ditching the extra wrapper script :).

**Comment by NiLuJe** [comment]
Speaking from experience on Kindles, one question I'm left with is what happens to inotify watches when onboard gets unmounted to be exported over USBMS?

**Comment by houqp** [comment]
> Speaking from experience on Kindles, one question I'm left with is what happens to inotify watches when onboard gets unmounted to be exported over USBMS?

We should probably reinitialize the watch every time it's mounted again...

**Comment by KenMaltby** [comment]
If this, like fmon, is dependent upon Nickel's processing of the .png, to put it in the db and library, then it retains the obstacle most new users fall prey to.  If you have totally automated the process, so it functions without requiring user intervention, that would make the install more reliable.

**Comment by NiLuJe** [comment]
What obstacle would that be? (never used fmon, sorry for rehashing old
stuff ;))

**Comment by KenMaltby** [comment]
Where users most often ran into problems, when using the install guides I posted at MR, was with getting the .png files properly processed for fmon's use.

**Comment by NiLuJe** [comment]
Guess we'll have to check if putting it in the KoboRoot tarball gets it
processed at boot?

**Comment by KenMaltby** [comment]
Nickel's processing is triggered upon return from a USB connection. This makes the monitoring process a two step install process for the user.

**Comment by NiLuJe** [comment]
The idea behind the SQL query would be to avoid such cases. (Note to self: for the same reason, check IN_OPEN events and not IN_CLOSE, to avoid catching the eventual CLOSE of nickel's processing, at which point the DB *may* already be updated, but we certainly don't want to start anything).
I'll double-check, but I don't think we trip anything when removing the PNG from Nickel?

I thought about adding the SQL entry ourselves, but that seems a bit overkill when nickel does it for us. And in any case, the only thing I can think of where we could run that code from on a vanilla device rerquires a reboot.

----

To get back to the install:
* Check if bundling the PNG in KoboRoot indeed gets it processed at boot...
* Barring that, we've had good experience on KIndles to ship  a zip file with the full directory structure prepared, with a simple instruction of: unpack to the USB root. That would have the added benefit of unanonimyzing the KoboRoot tarball since it'd be inside that zip.

**Comment by NiLuJe** [comment]
Deleting from the Library:

```
/mnt/onboard/koreader.png ATTRIB 
/mnt/onboard/koreader.png DELETE_SELF
```

So I'm not quite sure how one could get "trapped", @KenMaltby, do you have real-world examples?

**Comment by pazos** [comment]
Whoooaa! @NiLuJe , when do you sleep?

The loop that @KenMaltby refers is found if fmon is active before koreader.png is indexed. This is why I think about checking nickel db in first place.

@houqp : we don't need to reinitialize the watch. If file isn't found there isn't a way to open the file. The program will run even if /mnt/onboard/koreader.png doesn't exist

**Comment by NiLuJe** [comment]
@pazos: Not much tonight :D.

As for the issue with reinitializing watch, the watch automatically gets destroyed when the underlying FS is unmounted (which is what should happen when it gets exported over USBMS).

fmon doesn't have any code to deal with this, so I'm wondering if this ever was an issue, or if Kobo somehow handles the USBMS thing differently than what I expected?

**Comment by houqp** [comment]
In this case, sounds like it should be a trivia fix? i.e. just wait until it's indexed in the db before performing a restart.

sqlite uses flock, so it's safe for multiple processes to write to it at the same time: https://www.sqlite.org/faq.html#q5. That said, I would still avoid doing it because I don't really trust sqlite ;P

**Comment by NiLuJe** [comment]
@houqp: I'm more concerned about further USBMS events down the line...

I'll check if fmon implemented a workaround it the easy way (i.e., hook into something else relating to mounts and simply restart the fmon process when onboard is ready). (Which, incidentally, is how I went about it on legacy Kindles when using inotify-tools ;p).

Barring that, there are ways to wait for FS events, so, look into that as well. (Which is what i'd prefer anyway, centralizing everything in a single C daemon, without any wrapper/monitoring shell scripts involved).

**Comment by NiLuJe** [comment]
Okay, I have all the info I need... I'm going to take a stab at this.

Prepare to be amazed by how terrible I am at C :D.

**Comment by NiLuJe** [comment]
Okay, WIP in NiLuJe/kfmon ...

Should currently handle unmounts/remounts properly!
Going to tackle SQLite next :).

**Comment by NiLuJe** [comment]
And now with added SQL! :)

**Comment by NiLuJe** [comment]
And it's now a true little daemon :).

**Comment by NiLuJe** [comment]
Okay, did a testrun on a Kobo.

It basically works as intended and as well as it did on my local box when playing with a bunch of tmpfs ;p.

The only major issue left is what happens when opening koreader.png from inside KOReader... It's... not good.
Since we block on system(), nothing happens while KOReader is up... but as soon as you exit it, nickel restarts... but so does KOReader. And then you're basically screwed (I had to dig out a toothpick for the reset button).

So far my less sucky idea is twofold:
* Background the call to koreader.sh in the system() call (that way we should not be blocking anymore, which helps us avoid the catastrophic timing of KOReader exit -> Nickel STartup + KOReader startup...
* Update the koreader script to prevent lauching concurrent instances of itself.

**Comment by NiLuJe** [comment]
But first, it's sammich time!

**Comment by NiLuJe** [comment]
Ooookay, after battling execve() & friends, I think I have something that won't try to launch seventy KOReaders concurrently, and won't leave zombie processes either.

**Comment by NiLuJe** [comment]
Okay, this one appears to be a winner :).

Checking the KoboRoot now...

**Comment by NiLuJe** [comment]
Okay, because that would have been too easy, putting stuff in onboard from a KoboRoot doesn't process it...

**Comment by KenMaltby** [comment]
There is a way to simulate a USB connection and initiate the Nickel processing.

**Comment by NiLuJe** [comment]
Nah, the zip approach works ;).

It processes everything before rebooting to install the KoboRoot :).

**Comment by NiLuJe** [comment]
... And on that note, I can confirm that writing to the DB while nickel holds it has a high probability of trashing it xD.

I was testing a bit of code that updated the Title/Author/Comment of our icon in the Library, it went fine for 5 or 6 tries... And then I got the dreaded error thing plus forced signout.

So, am going to disable that bit of code for now ;o).

**Comment by NiLuJe** [comment]
And I just re-trashed it because I'm very stupid and re-tapped on koreader in a search result... :D.

**Comment by NiLuJe** [comment]
Hmmm. Now I'm thinking it's either nickel being sneaky, or the DB getting trashed because we're killing nickel slightly too savagely...

**Comment by NiLuJe** [comment]
Not any weird zombie checking, at the very least. And on the plus side, we now don't leave any hanging for any period of time, yay.

I'm thinking what went wrong (some) of these time is being triggered during the thumbnail creation. Should probably check the DB for a created thumbnail instead of only the book...

**Comment by houqp** [comment]
what do you mean by trashing the db? Did it got corrupted? When do we kill the nickel process?

BTW: awesome work on the kfmon daemon :)

**Comment by NiLuJe** [comment]
Well, I initially assumed it was getting corrupted (I get the "Error" popup asking to sign out). But the fact that, in addition to that, nickel is booting downscaled in the top-left corner of the screen might be a hint hat there's some more weirdness involved in how we try to restart nickel...

We kill nickel in the beginning of koreader.sh when it's up.

So, besides trying to figure it out over USBNet, one cheap thing to try would be stopping trying to be awesome (being able to restart nickel), and just reboot when exiting KOReader when started from KFMon... 

What's annoying is that it worked perfectly fine at one point... But I don't remember exactly how much crap changed between the last time it worked and now... :D.
I remember that I originally started from the Homescreen tile or the Library, and the thumbnail was already processed, whereas I now launch it through a search for koreader (and it's actually the thumbnail creation that triggers KFMon -- couldn't find anyway to detect if that's already done from the DB itself... Which is weird, since those must be stored *somewhere*...).
But then, it might also only be broken since the switch from system() to execvp()...

>_<"

**Comment by NiLuJe** [comment]
Okay, I really need to sleep. I *know* where the thumbnails are stored, I tinkered with that in Calibre last year...

I just have to remember where the crappy split path is stored...

**Comment by KenMaltby** [comment]
OT comment: Since this install provides a launching process that requires having Nickel running before the launch, of KOReader, can take place, I will not be using it.  I will stick with KSM as I can then launch KOReader without starting Nickel at all.  (AdvBoot, is also an option for those of us who would rather not have started Nickel to use KOReader.)  A "One Step Install" is a great advance for the majority of users, though.
Luck;
Ken

**Comment by NiLuJe** [comment]
Okay, I think I ironed things out...

Turns out it worked fine when I launched kfmon from an SSH shell (I think)...
After some more tweaks to the KOReader startup script to siphon the full environment from nickel, it appears to behave.

I added a bunch more safety checks in the meantime, so, there's that, too ;).

**Comment by NiLuJe** [comment]
And in any case, we do need to reboot, because even if restarting nickel now works, we're still left with the broken suspend behavior we've discussed and that I had completely forgotten ;D.

**Comment by NiLuJe** [comment]
Okay, so, if anyone wants to check it out...

Install package: [http://files.ak-team.com/niluje/kopub/KFMon.zip](Here!). Just unzip at the USB root :).

Note that it very, very, very much depends on #2001, so, update your KOReader startup script first!

And, of course, *BACKUP YOUR DATABASE & SETTINGS*! I'd go so far as backing up the full USB storage, but I'm pleased to say that I only have had to resort to DB & settings backups ;p.

**Comment by NiLuJe** [comment]
@houqp: If you could add a static SQLite build somewhere in third-party, I'd very much appreciate it! (I'm afraid of screwing things up w/ CMake...).

A few pointers:
  * configure w/ --enable-static --disable-shared --enable-threadsafe
  * Append -fno-ffast-math to CFLAGS

**Comment by NiLuJe** [comment]
And, in case of trouble, the logs are stored in ```/usr/local/kfmon/kfmon.log```, and should be verbose enough to help me pinpoint the issue ;).

**Comment by NiLuJe** [comment]
Also, it depends on FW >= 2.9.0 (and slightly possibly slightly more recent to avoid SQL issues, I'd have to roughly check the timeline).

**Comment by houqp** [comment]
> @houqp: If you could add a static SQLite build somewhere in third-party, I'd very much appreciate it! (I'm afraid of screwing things up w/ CMake...).

Sure, but you mean adding it to kfmon right? Unless you don't want to make it into a generic launcher, then maybe we can include it into koreader.

**Comment by NiLuJe** [comment]
@houqp: I'd like to avoid reinventing the wheel and somehow vendor SQLite in KFMon directly, and right now it's pretty KOReader specific, yeah.
So yeah, I was thinking simply build SQLite as part of our third-party stuff, and then adapt the KFMon Makefile to point to it for linking? But if that feels wonky I can find another way ;).

**Comment by NiLuJe** [comment]
Updated the install package with the latest round of fixes ;).

**Comment by pazos** [comment]
> And in any case, we do need to reboot on exit, because even if restarting nickel now works, we're still left with the broken suspend behavior we've discussed and that I had completely forgotten ;D.

@NiLuJe What behaviour?. I use fmon and can switch between nickel and koreader without broken suspend. (6% of battery loss on 3 days in koreader)

**Comment by NiLuJe** [comment]
@pazos: If things go wrong at any point, in inherits our broken suspend behavior and doesn't seem to be able to fix it (i.e., crazy looping trying to deal with the /sys/power/* stuff).
Whether that actually affects battery life in Nickel, I couldn't tell you, but it's certainly not pretty to watch over strace ;).

**Comment by pazos** [comment]
@NiLuJe: when you talk about "broken suspend behaviour", do you mean that it isn't battery friendly? Did I miss some related issues?

i ran my kobo aura hd for two months without a reboot. Used nickel to read pocket articles and KOReader to read epubs & pdfs. I used both sleep functions (via a magnet cover) without troubles. Had no problems at all.

I did notice that long sleep times seems to have less impact on nickel than on koreader. But AFAICT this can't be called broken suspend..

pd: sorry about my english!

**Comment by NiLuJe** [comment]
@pazos: I don't really know. Like I said, it's weird as hell and it's hard to actually tell what's happening and the impact it has. But something is certainly not going as planned, and that irks me ;p.

**Comment by houqp** [comment]
@NiLuJe there you go: https://github.com/koreader/koreader-base/pull/418.

**Comment by houqp** [comment]
@pazos what version of KOReader are you using? We might have broken the suspend feature in kobo recently...

**Comment by NiLuJe** [comment]
Because I'm a tiny bit crazy, I've now made KFMon completely user-configurable.

i.e., you can define your own file/action couples via a simple .ini file dropped in a config directory.

Haven't had the chance to actually test those on a real Kobo because Savannah's network is wonky, so my builds are stalled...

**Comment by pazos** [comment]
@houqp I'm on v2015.11-394-g7816c33 right now, but I update all the time and had zero suspend issues last two months (both battery drains or broken suspend and returning home). I'm actually using fmon to launch koreader.

@NiLuJe Woow, great work on kfmon. I like fancy user configuration. How about to watch a whole folder (like /mnt/onboard/kfmon) looking for some-fancy.png and run the related some-fancy.sh. This would make easier to add new programs to the launcher  and kfmon a good candidate for kobo general launcher.

Offtopic: I got a 2nd hand, software "broken", kobo aura hd for less than 40€ which is now in factory reset state and ready to test updates.

**Comment by NiLuJe** [comment]
@pazos: Technically feasible, but recursive inotify watches are iffy to handle, so, meh.
Would require a bit of refactoring, so, unlikely to happen.

It's basically three lines in a .ini to add a watch, that shouldn't be a huge barrier ;).

**Comment by pazos** [comment]
@NiLuJe : Finally tested!!!! Awesome work!!!  

Attached a comparation video between fmon and kfmon..
Both work really well for me but fmon is faster because koreader.sh don't force a reboot when fmon is detected.

[fmon vs kfmon.zip](https://github.com/koreader/koreader/files/265086/fmon.vs.kfmon.zip)

**Comment by NiLuJe** [comment]
Yeah, not rebooting *should* be safe, in general, but I took the ~5s hit as a safety precaution for now ;).

**Comment by NiLuJe** [comment]
Namely, switch [https://github.com/koreader/koreader/blob/master/platform/kobo/koreader.sh#L123](this) ```reboot``` to ```./nickel.sh &``` ;).

**Comment by NiLuJe** [comment]
I might make that user-configurable while I'm at it...

**Comment by KenMaltby** [comment]
Wouldn't the "return 0" get you back to the launching program?  The "reboot" seems unneeded, if things were handled properly when reader.lua was called.  As it is now when I use KOReader's Home Icon I return to KSM without rebooting.

**Comment by NiLuJe** [comment]
@KenMaltby: Indeed, except in this instance we have *nothing* to return to, since we basically kill everything at the beginning of the KOReader startup script ;).

Hence we either need to restart Nickel ourselves, or reboot.

**Comment by KenMaltby** [comment]
I see, but is a reboot needed to return to KSM or Advboot ?

**Comment by NiLuJe** [comment]
Not with KSM, since it's still running in the BG. Which is why we don't kill it and simply return to it.

**Comment by NiLuJe** [comment]
As for AdvBoot, it requires a reboot also.

**Comment by NiLuJe** [comment]
Okay, got a Kobo test build running, appears to behave properly...

Going to write some doc, tag a release, and push this to the world...

**Comment by NiLuJe** [comment]
Okay, done!

Updated the [install package](http://files.ak-team.com/niluje/kopub/KFMon.zip).
Update the [docs](https://github.com/NiLuJe/kfmon).
Comited my [KOReader script tweaks](https://github.com/NiLuJe/koreader/commit/892e3c0f1e5633d9fbc8626ff114d45bee101997).

I'm waiting for those to hit a nightly before pushing this to MR :).
In the meantime, feel free to test & comment :).

**Comment by KenMaltby** [comment]
Not to put a damper on the Great work @NiLuJe has accomplished in creating this "Kute File Monitor", improved fmon, But is there a one step installer for KOReader?  The new launcher looks neat, for those who use Nickel, but I was hoping that the automated installation of KOReader (mostly for newcomers) was being included in this issue.

**Comment by NiLuJe** [comment]
@KenMaltby: If you basically wrap a koreader zip in that install package, you've got it down to one step ;).

**Comment by houqp** [comment]
Awesome work on kfmon! Looking forward to the MR post. No worry about the sqlite dep, i think it makes more sense to vendor it inside kfmon :)

I assume you are going to send out a PR for the tweaked start script?

**Comment by houqp** [comment]
NOTE to myself or anyone who is interested to help maintaining the doc:

Update the installation guide in wiki after kfmon is officially launched.

**Comment by NiLuJe** [comment]
@houqp: Yep, I'm waiting on feedback from #2005 for the PR, since I have a mix of stuff relevant to that in my tree :p.

**Comment by KenMaltby** [comment]
The MR post will be a "Sticky", when I see it.

**Comment by NiLuJe** [comment]
[MR Thread](http://www.mobileread.com/forums/showthread.php?t=274231) posted ;).

As for this issue, I'm thinking maybe do a true 'one-step' package for the next KOReader *stable* release?
(Provided nothing drastically broken emerges on the KFMon side of things ;)).

**Comment by KenMaltby** [comment]
It looks like they have a preferred way to handle this sort of thing:

"Hi Ken,

 Thank you for your suggestion of a post to be made sticky. We would recommend that posts of this nature - instructions for doing something - be put in the MR Wiki rather than in the forums. The Wiki is a far better place for factual information of a fairly static nature such as this. It doesn't really belong in the forums, which are more suited for discussions.

 Kind regards,

 HarryT (for the MobileRead moderating team)"

Luck;
Ken

**Comment by NiLuJe** [comment]
Fair enough ;). I'm not overly fond of stickying everything either (in fact, it's sometimes proved to be counter-intuitive in the Kindle Dev subforum, where stickied stuff seems to go unnoticed).

Granted, the fact that a bunch of the active devs got mod powers in the forum help for this kind of boring quality of life stuff ;).

**Comment by pazos** [comment]
@NiLuJe @houqp : Sooo, enhancement done!. Should I close the issue or wait for documentation updates on https://github.com/koreader/koreader/blob/master/platform/kobo/fmon/README.txt ?

I think installation based on fmon should be deprecated and documentation moved elsewhere. We can make a lean zip without Koboroot.tgz, koreader.png...

**Comment by KenMaltby** [comment]
@pazos The first ten lines should be retained, (with the references to the old method removed). They could be moved to follow the new KFMon method, though.

**Comment by houqp** [comment]
yeah, we should update that README to replace fmon with kfmon. BTW, why is that README in fmon directory? Shouldn't it be moved to its parent dir?

**Comment by Frenzie** [comment]
Also note that the text has acquired a few discrepancies with the wiki: https://github.com/koreader/koreader/wiki/Installation-on-Kobo-devices

**Comment by KenMaltby** [comment]
I think I may be falling behind, So will KOReader, for Kobo, now come with its own launcher?  Will it be included in the download install package? Then the download package would include all that is needed to install and use KOReader, no need to download and install a launcher program? I take it then that the install process would install KFMon and copy over a "koreader folder" ?  

For those of us who would prefer to use KSM, will KFM coexist with KSM?  fmon does, (at least if you have an almost empty DB) and KSM makes use of a fmon process for some of its features.  If not; would just disabling or removing the KoboRoot.tgz, from the .zip, still allow the install?

**Comment by NiLuJe** [comment]
KFMon is completely orthogonal to KSM, one doesn't care about the other, and vice-versa. As long as you don't open any of the trigger files in KSM, nothing happens.

Nothing is likely to change on KOReader's side. I'll probably just make a one-step zip available *somewhere* of the next stable release, and that's it.

**Comment by pazos** [comment]
@NiLuJe :  bug found!

Changing from rebooting to restarting does not work. Nickel won't launch (stuck in animation)

log said:

`
[KFMon] [2016-05-22 @ 00:22:53] Trying to load config file '/mnt/onboard/.adds/kfmon/config/kfmon.ini' . . .
[KFMon] [2016-05-22 @ 00:22:53] Daemon config loaded from 'kfmon.ini': db_timeout=450, use_syslog=0
[KFMon] [2016-05-22 @ 00:22:53] Beginning the main loop.
[KFMon] [2016-05-22 @ 00:22:53] Initializing inotify.
[KFMon] [2016-05-22 @ 00:22:53] Setup an inotify watch for '/mnt/onboard/koreader.png' @ index 0.
[KFMon] [2016-05-22 @ 00:22:53] Listening for events.
[KFMon] [2016-05-22 @ 00:24:00] Initializing KFMon v0.9-16-g9b914a8 | Built on May 16 2016 @ 23:24:16 | Using SQLite 3.12.2 (built against version 3.12.2)
[KFMon] [2016-05-22 @ 00:24:00] Trying to load config file '/mnt/onboard/.adds/kfmon/config/koreader.ini' . . .
[KFMon] [2016-05-22 @ 00:24:00] Watch config @ index 0 loaded from 'koreader.ini': filename=/mnt/onboard/koreader.png, action=/mnt/onboard/.adds/kfmon/my_koreader.sh, do_db_update=0, db_title=KOReader, db_author=KOReader Devs, db_comment=An eBook reader application
[KFMon] [2016-05-22 @ 00:24:00] Trying to load config file '/mnt/onboard/.adds/kfmon/config/kfmon.ini' . . .
[KFMon] [2016-05-22 @ 00:24:00] Daemon config loaded from 'kfmon.ini': db_timeout=450, use_syslog=0
[KFMon] [2016-05-22 @ 00:24:00] Beginning the main loop.
[KFMon] [2016-05-22 @ 00:24:00] Initializing inotify.
[KFMon] [2016-05-22 @ 00:24:00] Setup an inotify watch for '/mnt/onboard/koreader.png' @ index 0.
[KFMon] [2016-05-22 @ 00:24:00] Listening for events.
[KFMon] [2016-05-22 @ 00:24:14] Tripped IN_OPEN for /mnt/onboard/koreader.png
[KFMon] [2016-05-22 @ 00:24:14] Tripped IN_CLOSE for /mnt/onboard/koreader.png
[KFMon] [2016-05-22 @ 00:24:14] Spawning /mnt/onboard/.adds/kfmon/my_koreader.sh . . .
[KFMon] [2016-05-22 @ 00:24:14] . . . with pid: 895

`

**Comment by NiLuJe** [comment]
@pazos: I assume you're of course running a

**Comment by pazos** [comment]
@NiLuJe:

Yes, lastest nightly

With old kfmon (the one without .ini support) and old koreader.sh (without .ini logic to decide between launch nickel and reboot) changing `reboot` to `./nickel.sh &` in koreader.sh did the work.

With stock device + your tools + lastest kfmon + lastest koreader restarting don't work. Reboot is fine. 

I have no time right now to look at it so I changed back to reboot. Everything worked perfect here.

**Comment by NiLuJe** [comment]
Mmh, thanks for the details, I'll look into it... I probably did something
stupid somewhere along the way ;).

**Comment by NiLuJe** [comment]
Tiny bit of necromancy to mention that I fixed [KFMon](https://github.com/NiLuJe/KFMon), and it now deals with nickel restarts properly (along a host of other bugfixes/improvements) ;p.

(Better late than never!).

**Comment by NiLuJe** [comment]
And two years later, [one-step](https://www.mobileread.com/forums/showthread.php?t=314220) packages for real! :D.
